### PR TITLE
beginHUD/endHUD projection fix

### PR DESCRIPTION
**Pull request by diwi**
beginHUD/endHUD seems to fail for custom set camera projections, e.g. a FoV different than 60 degrees.
I just found this by accident and am a bit puzzled that I never encountered it before.

`
perspective(80 * DEG_TO_RAD, width/(float)height, 1, 5000);`

Before:
The rectangular border should align with the screen-border.
Note, lightning also effects the shading of the HUD-scope.
bug_hud

After:
fixed_hud

Example (also in the PR):

```java

import peasy.*;

// camera control
PeasyCam peasycam;

PGraphics2D pg_screen;

public void settings() {
  size(800, 600, P3D);
  smooth(8);
}

public void setup() {
//    surface.setResizable(true);
  
  // default FoV is 60
  perspective(80 * DEG_TO_RAD, width/(float)height, 1, 5000);

  // camera
  peasycam = new PeasyCam(this, 300);
  
  
  // just some mask for HUD display test
  pg_screen = (PGraphics2D)createGraphics(width, height, P2D);
  pg_screen.smooth(0);
  pg_screen.beginDraw();
  {
      pg_screen.clear();
      pg_screen.beginShape();
      pg_screen.fill(0xFF000000); pg_screen.vertex(0, 0);
      pg_screen.fill(0xFFFF0000); pg_screen.vertex(width, 0);
      pg_screen.fill(0xFFFFFFFF); pg_screen.vertex(width, height);
      pg_screen.fill(0xFF00FF00); pg_screen.vertex(0, height);
      pg_screen.endShape();
      pg_screen.blendMode(REPLACE);
      pg_screen.noStroke();
      pg_screen.fill(0,0);
      pg_screen.rect(20, 20, width-40, height-40);
  }
  pg_screen.endDraw();
}

public void draw(){
  
  // in case of surface resizing (happens asynchronous) this has no effect in setup
  // perspective(120 * DEG_TO_RAD, width/(float)height, 1, 5000);
  // peasycam.feed();
  
  
  // 3D scene
  ambientLight(128, 128, 128);
  pointLight(255, 128, 64, -200, -200, 10);
  pointLight(64, 128, 255, +200, +200, 10);
  
  background(16);
  
  rectMode(CENTER);
  strokeWeight(1);
  stroke(0);
  fill(128);
  rect(0, 0, 400, 400);
  
  strokeWeight(2);
  stroke(255, 64,  0); line(0,0,0,100,0,0);
  stroke( 32,255, 32); line(0,0,0,0,100,0);
  stroke(  0, 64,255); line(0,0,0,0,0,100);
  
  translate(80,80,80);
  noStroke();
  fill(128);
  box(50);
  
  
  // screen-aligned 2D HUD
  peasycam.beginHUD();
  image(pg_screen, 0, 0);
  peasycam.endHUD();

}

```

### CODE REVIEW

**Comment by jdf** [comment]
Hi. I'm interested in this work, but right now there's a lot of commented lines, making it hard to understand the changes. It looks like you're still in the middle of working on it. Let me know when you're ready for me to review it.

**Comment by diwi** [comment]
The comments are the original code. Didn't want to get rid of it just yet.

**Comment by diwi** [comment]
But basically i think thats it.

I'm not sure about  `g.noLights();` though.
It is useful, but it is not mandatory for the beginHUD/endHUD scope to work.

**Comment by jdf** [comment]
When you're 100% ready for review (meaning, it 100% works, and there's no old code lying around), let me know.

**Comment by diwi** [comment]
Hi,
cleaned up and ready to review.

File: src/peasy/PeasyCam.java [hunk 1]
```diff
@@ -581,21 +579,53 @@ public void setRotations(final double pitch, final double yaw, final double roll
 		return new float[] { 0, 0, 0 };
 	}
 
+	
+	
+	
+	private boolean pushed_lights = false;
```
Comment (jdf) [inline src/peasy/PeasyCam.java]: savedLights or pushedLights

File: src/peasy/PeasyCam.java [hunk 2]
```diff
@@ -581,21 +579,53 @@ public void setRotations(final double pitch, final double yaw, final double roll
...
+	 * end screen-aligned 2D-drawing.
+	 * 
+	 */
 	public void endHUD() {
-		g.hint(PConstants.ENABLE_DEPTH_TEST);
+		if(g.isGL()){
+			PGraphicsOpenGL pgl = (PGraphicsOpenGL)g;
+			pgl.popProjection();
+			pgl.lights = pushed_lights;
+			pushed_lights = false;
```
Comment (jdf) [inline src/peasy/PeasyCam.java]: This line isn't necessary.

File: src/peasy/PeasyCam.java [hunk 3]
```diff
@@ -581,21 +579,53 @@ public void setRotations(final double pitch, final double yaw, final double roll
...
-		// Apply the original Processing transformation matrix.
-		g.applyMatrix(originalMatrix);
+		if(g.isGL()){
+			PGraphicsOpenGL pgl = (PGraphicsOpenGL)g;
+			pgl.pushProjection();
+			pushed_lights = pgl.lights;
+			pgl.lights = false;
+		}
+		if(g.is3D()){
+			g.ortho(0, g.width, -g.height, 0, -Float.MAX_VALUE, +Float.MAX_VALUE);
```
Comment (jdf) [inline src/peasy/PeasyCam.java]: Does this not need to be saved/restored?

**Comment by diwi** [comment]
"All checks have failed" ... not sure why this keeps popping up?

checks_failed

**Comment by jdf** [comment]
Ignore it.

**Comment by diwi** [comment]
I think it's all covered now. Are there further things to look for?

File: src/peasy/PeasyCam.java (hunk 3 above)
Comment (jdf) [reply src/peasy/PeasyCam.java]: Still need an answer to this question.

Comment (diwi) [reply src/peasy/PeasyCam.java]: I'n not exactly sure what you are referring too. All states are saved restored.

Comment (diwi) [reply src/peasy/PeasyCam.java]: Checked a couple of times now and cant see whats missing:

https://github.com/diwi/peasycam/blob/6c069f210ee7709cb0dbdd91b240f6ae60792700/src/peasy/PeasyCam.java#L599-L628

**Comment by jdf** [comment]
What is the setting that corresponds to undoing the `ortho` call?

**Comment by jdf** [comment]
Thank you.

**Comment by jdf** [comment]
woohoo! I'll build and deploy.

**Pull request merged**